beanie/odm/utils/encoder.py

```diff
@@ -0,0 +1,122 @@
+from collections import defaultdict
+from datetime import datetime
+from enum import Enum
+from pathlib import PurePath
+from types import GeneratorType
+from typing import (
+    AbstractSet,
+    Any,
+    Callable,
+    Dict,
+    List,
+    Mapping,
+    Set,
+    Tuple,
+    Union,
+)
+from uuid import UUID
+
+from pydantic import BaseModel
+
+from bson import ObjectId
+
+from .bson import ENCODERS_BY_TYPE
+
+SetIntStr = Set[Union[int, str]]
+DictIntStrAny = Dict[Union[int, str], Any]
+
+
+def generate_encoders_by_class_tuples(
+    type_encoder_map: Dict[Any, Callable[[Any], Any]]
+) -> Dict[Callable[[Any], Any], Tuple[Any, ...]]:
+    encoders_by_class_tuples: Dict[
+        Callable[[Any], Any], Tuple[Any, ...]
+    ] = defaultdict(tuple)
+    for type_, encoder in type_encoder_map.items():
+        encoders_by_class_tuples[encoder] += (type_,)
+    return encoders_by_class_tuples
+
+
+encoders_by_class_tuples = generate_encoders_by_class_tuples(ENCODERS_BY_TYPE)
+
+
+def bsonable_encoder(
+    obj: Any,
+    exclude: Union[
+        AbstractSet[Union[str, int]], Mapping[Union[str, int], Any], None
+    ] = None,
+    by_alias: bool = True,
+    custom_encoder: Dict[Any, Callable[[Any], Any]] = {},
+) -> Any:

```

Comment:
I think it will return not any, but Mapping of something or Dict of smth. If it is hard to say now, what will be the result, pls add TODO flag here. The same about `obj`

Reply:
`bsonable_encoder` can accept any type you pass in it and return any type mappend to that type.
It is called recursively on every field except dict keys.

Reply:
I see. Thank you for the clarification